### PR TITLE
initial support for pmr containers in DPL

### PR DESCRIPTION
**Pull request by mkrzewic**
depends on #1581

### CODE REVIEW

**Comment by mkrzewic** [comment]
@ktf, indeed, the need for include was already gone, forgot to change that back. done.

**Comment by mkrzewic** [comment]
@ktf on another note - according to @rbx the boost includes are gone - is that verified?

**Comment by ktf** [comment]
I had a bit of time to rethink about this. I think you have been too conservative. ;-) IMHO, what we should do is to replace newChunk and adoptChunk to basically be what you do in `makeVector` / `adoptVector`, basically having `std::pmr::vector` replace the `DataChunk`. We can then discuss how to expose a proper API for all the other types, but given that FairMQ now basically supports resizable messages I do not see why DataChunk should limit us to fixed size... I would still not expose the container to the user, like in the current interface, but I guess that for the backend there is not much to gain from having it based around fixed size messages no?

@matthiasrichter what do you think?

**Comment by mkrzewic** [comment]
AFAIK there is no support for resizeable messages in FairMQ - last I checked you could only specify a part of the allocated buffer to be sent, the full resize functionality lies with the container (@rbx something new was added?).
Anyway the point of this PR is exactly to expose a container and skip the "ownership manager" abstraction layer, there is no need for that in modern C++, it just makes things complicated.
I consider the "adoptContainer" vital functionality. The "makeVector" part is just for convenience introduced at @ktf request, not really necessary.
I think this PR does not break anything in DPL, so I don't see why it takes so long to merge.

**Comment by davidrohr** [comment]
@ktf @mkrzewic : This has been stalled since >3 months now. Do we still need it, then we should rebase and merge it at some point?

**Comment by ktf** [comment]
I would like to see the developments of @matthiasrichter in the same direction before deciding what to do with this.

**Comment by matthiasrichter** [comment]
I think its independent of the other development which is basically an extension of #1873 / #1846. Its a matter of two different approaches and the arguments have been posted here in this PR.

**Comment by davidrohr** [comment]
Well, if it is independent and provides useful functionality, I don't see a reason not to merge it, in particular since it is a rather small change and looks elegant. But to be honest, I didn't dive deep enough into the framework to judge if this is just a competing strategy to #1846.

**Comment by sawenzel** [comment]
Is there any new development here? Probably a decision needs to be taken at some point.

**Comment by mkrzewic** [comment]
Dear all, i updated the PR to use Matthias' new MessageContext thingy thereby improving compliance with the way-DPL-does-it. As a reminder: this only *adds* functionality without breaking or changing anything (except for fixing the broken DPLMerger which looks like dead code anyway).

File: Framework/Utils/include/DPLUtils/Utils.h [hunk 1]
```diff
@@ -18,16 +18,14 @@
...
+#include "MemoryResources/MemoryResources.h"
 
 namespace o2f = o2::framework;
 
 namespace o2
 {
 namespace workflows
 {
-//TODO: this is to make DPLmerger compile, but this code is (and always was) horribly broken - please remove.
-inline void freefn(void* data, void* /*hint*/) { delete static_cast<char*>(data); };
```
Comment (davidrohr) [inline Framework/Utils/include/DPLUtils/Utils.h]: Difficult to judge for me, but if this is broken and should be removed, and it compiles after removal, then it is probably the way to go.

Comment (mkrzewic) [reply Framework/Utils/include/DPLUtils/Utils.h]: probably, a remnant after removal of the Stack::freefn long time ago.

**Comment by davidrohr** [comment]
@ktf @sawenzel : Could you have a look?

**Comment by ktf** [comment]
Ok, fine with me at this point. I still disagree on the idea, but I guess we need something. @matthiasrichter what do you think?

**Comment by ktf** [comment]
Yes, indeed, documentation and a test would be nice. I assume there is no way to simply call the two adopt and make?

**Comment by ktf** [comment]
Actually makeVector could become `make<o2::vector<T>>` and be folded as one of the `constexpr` ifs, no?

**Comment by mkrzewic** [comment]
@ktf, indeed, looks like it could. About the test: I'll think about a test workflow, not sure how to test this "standalone" to actually trigger all the involved dpl internals...

**Comment by ktf** [comment]
@mkrzewic there is a test already for DataAllocator:

https://github.com/AliceO2Group/AliceO2/blob/dev/Framework/Core/test/test_DataAllocator.cxx

Something simple there would do.

For the renaming we can do it afterwards, unless you find it completely trivial to do.

**Pull request merged**